x/dymns/keeper/dym_name.go

```diff
@@ -0,0 +1,800 @@
+// This function will remove the Dym-Name record as well as the reverse mappings records.
+func (k Keeper) DeleteDymName(ctx sdk.Context, name string) error {
+	if err := k.BeforeDymNameOwnerChanged(ctx, name); err != nil {
+		return err
+	}
+
+	if err := k.BeforeDymNameConfigChanged(ctx, name); err != nil {
+		return err
+	}
+
+	store := ctx.KVStore(k.storeKey)
+	dymNameKey := dymnstypes.DymNameKey(name)
+	store.Delete(dymNameKey)
+
+	return nil
+}
+
+// GetAllNonExpiredDymNames returns all non-expired Dym-Names from the KVStore.
+// This function will filter out all expired Dym-Names based on the time of the context.
+func (k Keeper) GetAllNonExpiredDymNames(ctx sdk.Context) (list []dymnstypes.DymName) {
+	store := ctx.KVStore(k.storeKey)
+
+	iterator := sdk.KVStorePrefixIterator(store, dymnstypes.KeyPrefixDymName)
+	defer func() {
+		_ = iterator.Close()
+	}()
+
+	for ; iterator.Valid(); iterator.Next() {
+		var dymName dymnstypes.DymName
+		k.cdc.MustUnmarshal(iterator.Value(), &dymName)
+
+		if dymName.IsExpiredAtCtx(ctx) {
+			continue
+		}
+
+		list = append(list, dymName)
+	}
+
+	return list
+}
+
+// GetAllDymNames returns all Dym-Names from the KVStore.
+// No filter applied, to be used in genesis export.
+func (k Keeper) GetAllDymNames(ctx sdk.Context) (list []dymnstypes.DymName) {
+	store := ctx.KVStore(k.storeKey)
+
+	iterator := sdk.KVStorePrefixIterator(store, dymnstypes.KeyPrefixDymName)
+	defer func() {
+		_ = iterator.Close()
+	}()
+
+	for ; iterator.Valid(); iterator.Next() {
+		var dymName dymnstypes.DymName
+		k.cdc.MustUnmarshal(iterator.Value(), &dymName)
+		list = append(list, dymName)
+	}
+
+	return list
+}
+
+// PruneDymName removes a Dym-Name from the KVStore, as well as all related records.
+func (k Keeper) PruneDymName(ctx sdk.Context, name string) error {
+	// remove SO (force, ignore active SO)
+	k.DeleteSellOrder(ctx, name, dymnstypes.NameOrder)
+
+	// remove historical SO
+	k.DeleteHistoricalSellOrders(ctx, name, dymnstypes.NameOrder)
+	k.SetMinExpiryHistoricalSellOrder(ctx, name, dymnstypes.NameOrder, 0)
+
+	dymName := k.GetDymName(ctx, name)
+	if dymName == nil {
+		return nil
+	}
+
+	// remove config
+	// This seems not necessary because we are going to remove the record anyway,
+	// but just let it here to clear the business logic
+	dymName.Configs = nil   // all configuration should be removed
+	dymName.Owner = ""      // no one owns it anyone
+	dymName.Controller = "" // no one controls it anyone
+
+	return k.DeleteDymName(ctx, name)
+}
+
+// ResolveByDymNameAddress resolves a Dym-Name-Address into an output address.
+//
+// For example:
+//   - "my-name@dym" => "dym1a..."
+//   - "another.my-name@dym" => "dym1b..."
+//   - "my-name@nim" => "nim1..."
+//   - (extra format) "0x1234...6789@nim" => "nim1..."
+//   - (extra format) "dym1a...@nim" => "nim1..."
+func (k Keeper) ResolveByDymNameAddress(ctx sdk.Context, dymNameAddress string) (outputAddress string, err error) {
```

Comment:
I already extracted a part of this method to a smaller func

Reply:
the method `ReverseResolveDymNameAddress` is one of the two I don't want to break down because it has the most complicated business logic.
I want to put it flattened as much as possible here so readers can read it more easy.
Anyway, I already extracted to smaller parts.